prov/gni/src/gnix_nic.c

```diff
@@ -333,30 +338,38 @@ static int __process_rx_cqe(struct gnix_nic *nic, gni_cq_entry_t cqe)
 	struct gnix_vc *vc;
 
 	vc_id =  GNI_CQ_GET_INST_ID(cqe);
-	vc = __gnix_nic_elem_by_rem_id(nic, vc_id);
 
-#if 1 /* Process RX inline with arrival of an RX CQE. */
-	if (unlikely(vc->conn_state != GNIX_VC_CONNECTED)) {
-		GNIX_DEBUG(FI_LOG_EP_DATA,
-			  "Scheduling VC for RX processing (%p)\n",
-			  vc);
-		ret = _gnix_vc_rx_schedule(vc);
-		assert(ret == FI_SUCCESS);
-	} else {
-		GNIX_DEBUG(FI_LOG_EP_DATA,
-			  "Processing VC RX (%p)\n",
-			  vc);
-		ret = _gnix_vc_dequeue_smsg(vc);
-		if (ret != FI_SUCCESS) {
-			GNIX_WARN(FI_LOG_EP_DATA,
+	/*
+	 * its possible this vc has been destroyed, so may get NULL
+	 * back.
+	 */
+
+	vc = __gnix_nic_elem_by_rem_id(nic, vc_id);
+	if (vc != NULL) {
```

Comment:
Yes.  Because there is a window of time when using auto progress that a thread may be going through the progress engine while one of the application threads is actively tearing down and endpoint (and hence its associated VCs)  before the rem_id for the vc is removed from the vector.  This check alone isn't sufficient to avoid problems surrounding this potential race condition.